src/python/gem5/components/cachehierarchies/ruby/caches/viper/corepair_cache.py

```diff
@@ -0,0 +1,124 @@
+# Copyright (c) 2024 Advanced Micro Devices, Inc.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice,
+# this list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+# this list of conditions and the following disclaimer in the documentation
+# and/or other materials provided with the distribution.
+#
+# 3. Neither the name of the copyright holder nor the names of its
+# contributors may be used to endorse or promote products derived from this
+# software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import math
+
+from m5.objects import (
+    CorePair_Controller,
+    MessageBuffer,
+    RubyCache,
+    TreePLRURP,
+)
+
+from gem5.components.processors.abstract_core import AbstractCore
+
+
+class CorePairCache(CorePair_Controller):
+    def __init__(
+        self,
+        l1i_size: str,
+        l1i_assoc: int,
+        l1d_size: str,
+        l1d_assoc: int,
+        l2_size: str,
+        l2_assoc: int,
+        network,
+        cache_line_size,
+        core: AbstractCore,
+    ):
+        """Creating CorePair cache controller. Consist of both instruction
+        and data cache for a pair of L1s and a single L2 cache shared between
+        them.
+        """
+        super().__init__()
+
+        self.send_evictions = core.requires_send_evicts()
+
+        self.L1Icache = RubyCache(
+            size=l1i_size,
+            assoc=l1i_assoc,
+            replacement_policy=TreePLRURP(),
```

Comment:
I don't know what the right balance is between configurability and needing to write your own. It seems the goal of stdlib is to get everyone to explicitly change these values by writing their own script. As far as I know you can change these values at any point before calling simulate (which calls instantiate).  This applies to your other comments below on register file type/size.

Reply:
Would like to hear from @powerjg on this then ... at minimum so I can understand as it seems like we'd want these things in the base file but perhaps not?

Reply:
I am thinking a good middle ground might be to add `kwargs` and do a check so that only specific values can be set. This would act similar to the current configs where we pass around `args` or `options` from `argparse`. Except with a "checked" kwargs dict we could: 
 - Have a default set of key/values so that:
    - It is not *required* to be passed to the stdlib class if you just want the defaults.
    - Only contains keys that are actually used which eliminates the problem with current command line arguments where you don't even know if the command line option does anything without comparing `config.ini`s.
 - Would not require having argparse at the top of every python script which clutters the system setup.

Or maybe just import some argparser from stdlib with all the accepted command line arguments and passing around `args` could accomplish the same goal with less effort.

Reply:
Just to follow up on this (a month later, sorry...). We should make this have a reasonable default, but not allow "parameterization." If your research is interested in changing the replacement policy, then you should create a new class that enables that parameter. *We cannot have a parameter for everything, everywhere, all the time.* It is not possible to make code like that maintainable.

The purpose of the stdlib is twofold:
1. Have components with reasonable defaults. (This is a work in progress, but the prebuilt boards like this one is a good example.)
2. Enable people to more explicitly and with less room for error specialize the components they are using.

I admit this is a change from the way many people "grew up" with gem5, but from a software engineering perspective, it's going to be good for the project in the long run.